cli/inspector.rs

```diff
@@ -4,546 +4,746 @@
-    &self.client
-  }
+#[derive(Default)]
+struct InspectorFlags {
+  waiting_for_session: bool,
+  session_handshake_done: bool,
+  on_pause: bool,
+}
 
-  fn base_mut(&mut self) -> &mut v8::inspector::V8InspectorClientBase {
-    &mut self.client
+impl InspectorFlags {
+  fn new(waiting_for_session: bool) -> RefCell<Self> {
+    let self_ = Self {
+      waiting_for_session,
+      ..Default::default()
+    };
+    RefCell::new(self_)
   }
+}
 
-  fn run_message_loop_on_pause(&mut self, context_group_id: i32) {
-    assert_eq!(context_group_id, CONTEXT_GROUP_ID);
-    assert!(!self.paused);
-    self.paused = true;
-
-    // Creating a new executor and calling block_on generally causes a panic.
-    // In this case it works because the outer executor is provided by tokio
-    // and the one created here comes from the 'futures' crate, and they don't
-    // see each other.
-    let dispatch_messages_while_paused =
-      future::poll_fn(|cx| match self.poll_unpin(cx) {
-        Poll::Pending if self.paused => Poll::Pending,
-        _ => Poll::Ready(()),
-      });
-    executor::block_on(dispatch_messages_while_paused);
-  }
+struct InspectorSessions {
+  new_incoming:
+    Pin<Box<dyn Stream<Item = Box<DenoInspectorSession>> + 'static>>,
+  handshake: Option<Box<DenoInspectorSession>>,
+  established: FuturesUnordered<Box<DenoInspectorSession>>,
+}
 
-  fn quit_message_loop_on_pause(&mut self) {
-    self.paused = false;
+impl InspectorSessions {
+  fn new(
+    inspector: *mut DenoInspector,
+    new_websocket_rx: UnboundedReceiver<WebSocket>,
+  ) -> RefCell<Self> {
+    let v8_inspector = unsafe { &mut **inspector };
+    let self_ = Self {
+      new_incoming: new_websocket_rx
+        .map(move |ws| DenoInspectorSession::new(v8_inspector, ws))
+        .boxed_local(),
+      handshake: None,
+      established: FuturesUnordered::new(),
+    };
+    RefCell::new(self_)
   }
 
-  fn run_if_waiting_for_debugger(&mut self, context_group_id: i32) {
-    assert_eq!(context_group_id, CONTEXT_GROUP_ID);
+  fn clear(&mut self) {
+    let _ = replace(
+      self,
+      Self {
+        new_incoming: stream::empty().boxed_local(),
+        handshake: None,
+        established: FuturesUnordered::new(),
+      },
+    );
   }
 }
 
-#[derive(Clone)]
-struct DenoInspectorHandle {
-  deno_inspector_ptr: *mut c_void,
+struct InspectorWakerInner {
+  poll_state: PollState,
+  task_waker: Option<task::Waker>,
+  parked_thread: Option<thread::Thread>,
+  inspector_ptr: Option<NonNull<DenoInspector>>,
   isolate_handle: v8::IsolateHandle,
-  interrupted: Arc<AtomicBool>,
 }
 
-impl DenoInspectorHandle {
-  pub fn new(
-    deno_inspector: &DenoInspector,
-    isolate_handle: v8::IsolateHandle,
-  ) -> Self {
-    Self {
-      deno_inspector_ptr: deno_inspector as *const DenoInspector
-        as *const c_void as *mut c_void,
+unsafe impl Send for InspectorWakerInner {}
+
+struct InspectorWaker(Mutex<InspectorWakerInner>);
```

Comment:
Why not also include the Arc here?

`struct InspectorWaker(Arc<Mutex<InspectorWakerInner>>);`

Also it would be very helpful to have a small docstring describing the purpose of this. Why doesn't `AtomicWaker` suffice?

Reply:
Maybe this is better called InterruptWaker ?